### ISSUE TITLE
Convert all PDFs and images in a directory to trusted RGBA

### ISSUE THREAD

**Issue by frozentime345** [QubesOS/qubes-issues#3966]
### Qubes OS version: R3.2

Sorry if this isn't the place to request features(a google search did me no good nor did qubes os faq). It might be nice to have convert all to trusted(pdfs/images) for a folder(ie for complex file structures).  

### Affected component(s): Feature Request

---

### General notes:

---

**Comment by sstt011** [comment]
@macdog345 What do you mean "convert all to trusted for a folder"?

**Comment by macdog345** [comment]
Like every pdf and every image in the folder would be converted. Rather than having to track down each pdf and image and right click it(if one had such a complex folder).

**Comment by sstt011** [comment]
@macdog345 Which program are you using eg `convert`? Are you doing this in an AppVM/TemplateVM/dom0?

**Comment by macdog345** [comment]
Im using the convert to trusted pdf function of qubes os and in an AppVM? is there some kind of command for this cuz i guess I could write a bash script to accomplish the above(or someone more bash skilled could =p).

**Comment by marmarek** [comment]
qvm-convert-pdf, qvm-convert-img, both accept only one file at a time.

**Comment by macdog345** [comment]
kk thanks! was this to right place to ask for feature quests(just for future ref)?

**Comment by andrewdavidwong** [comment]
> was this the right place to ask for feature requests(just for future ref)?

Yes. :slightly_smiling_face:

**Comment by sstt011** [comment]
@marmarek I'm still confused about what needs to be fixed. Is the issue to be able to convert multiple images at a time?

**Comment by macdog345** [comment]
Yes and no, you can convert multiple pdfs from my experience by selecting each pdf and nothing else, but right clicking a folder or selecting control A on a folder with various file types leaves the option unavailable. So just a minor issue but it would be nice to one day convert entire USBS and such to trusted.

**Comment by johnnyboy-3** [comment]
**The problem you're addressing (if any)**
If you want to convert multiple jpegs to trusted jpegs (using nautilus, select files, right click -> Convert To Trusted Img: qubes opens a DVM, converts one single file, closes the DVM. Then takes the next file, opens a DVM, converts the jpeg, closes the DVM, etc ... if you have 40 files you want to convert, qubes fires up 40 DVMs in total which takes a long time.


**Describe the solution you'd like**
Fire up a DVM just once, convert all the jpegs, and close the DVM.


**Where is the value to a user, and who might that user be?**
Saving lots of time.



**Describe alternatives you've considered**
I could try to convert them myself which is probably not secure.


**Additional context**
Tested with current up to date Qubes R4.0.3. NOT tested with R4.1



**Relevant [documentation](https://www.qubes-os.org/doc/) you've consulted**
Searching for this problem in github and found this issue



**Related, [non-duplicate](https://www.qubes-os.org/doc/reporting-bugs/#new-issues-should-not-be-duplicates-of-existing-issues) issues**
This issue.

**Comment by donob4n** [comment]
Is there any consideration for fixing this? Should we use just one DVM if the user selects multiple files or a folder or should be two different ways for do it?

> qvm-convert-pdf, qvm-convert-img, both accept only one file at a time.

Should this remain this way?

**Comment by rootnoob** [comment]
Not to bump old threads, is this something you're still wanting for R4.1?

**Comment by frozentime345** [comment]
Yes, specifically I wanted to easily convert an entire folder/usb but johnnyboy-3's point about using a single DVM to save time is probably the most important change. Although I dont think file type should matter, if a user selects multiple pdfs or jpegs or whatever it should all be pulled to a single dvm then each pdf/image should be converted to trusted and returned.

**Comment by rustybird** [comment]
It should be conveyed to the user that such a batch conversion of multiple documents in a single DisposableVM has **weaker security** properties: Any malicious document could tamper with the visual content of the other documents in the batch. (Or at least of the ones that happen to be processed after it, assuming an implementation where each converted document is copied back to the source VM before the next one is converted.)

**Comment by ethnh** [comment]
not sure if this is an issue anymore I am able to select multiple PDF's in thunar and right click + convert and it will

**Comment by wondoketu** [comment]
Hey just reading this now. I've made a python script that runs `qvm-convert-pdf` on all the pdf's in a directory. It still uses a a new DVM for each conversion, but wouldn't using the same DVM for all of them be a problem? If one pdf is malicious, could it affect the conversions of everything else in that dvm?

Regardless, what I do to speed up the time is multithread the forking of DVM's for this. However, as you might guess, opening up 40 DVM's at once is a huge problem, so I use semaphore's to manage my resources. Here's a sample script.

```
import subprocess
from threading import Thread, Semaphore
# semaphore count based on your system constraint
pool_semaphore = Semaphore(5)
def thread_handler(target_pdf_file):
    with pool_semaphore:
        subprocess.check_output([
            "qvm-convert-pdf",
            target_pdf_file
        ])

thread_list = []
for target_file in LISTED_FILES:
    thread_list.append(Thread(target=thread_handler,args=(target_file,)))
for thread in thread_list:
    thread.start()
for thread in thread_list:
    thread.join()
```

You can adjust this code to find all pdf/image files and work accordingly without needing to manually do it yourself.

**Comment by Jayant-kernel** [comment]
@ben-grande @marmarek 
I would like to work on this 
Looking at extending CLI to accept multiple files/dirs
--batch flag with a security warning (single DVM = faster, weaker isolation)
update the nautilus handler too
planning a PR soon
